### PR TITLE
[XLA:GPU] Fix reversed dynamic-slice offset for offloaded scan residuals

### PR DESCRIPTION
**Pull request by zmelumian972**
#ai-generated

## Summary
- AnalyzeDynamicSlice substituted the loop counter into a called computation's parameter whenever `required_parameters` flagged it, but that flag only means the parameter is needed to reach the induction variable, not that it equals it.
- When the caller passes a function of the induction variable instead (e.g. `n-1-i`, which the transpose of a scan passes when reading a residual back), the slice was annotated with a forward address walk instead of a reverse one.
- This only surfaces for host offloading, since `HostMemoryTransferAsyncifier` is what moves such a slice into a called computation in the first place; device-memory slices stay inline in the while body and are analyzed correctly.
- Symptom: silently wrong gradients for a `jax.lax.scan` whose residuals are offloaded via `save_and_offload_only_these_names`, with no error and correct-looking memory figures.
- Fix: bail out (`return std::nullopt`) unless the caller forwards the induction variable unchanged.

## Test plan
- [x] Added `DynamicSliceAnalysisTest.ParameterIsFunctionOfInductionVariable` covering a called-computation parameter that receives `n-1-i` rather than the induction variable itself.
- [x] Verified against a standalone JAX repro (`jax.checkpoint` + `save_and_offload_only_these_names` over a `jax.lax.scan`): gradients matched a non-offloaded reference after the fix, and diverged (`max abs error ~770`) before it, reproduced on two different GPU architectures (Volta and Blackwell).

## Repro

Standalone JAX repro that demonstrates the bug end-to-end (`jax.checkpoint` + `save_and_offload_only_these_names` over a `jax.lax.scan`, compared against a non-offloaded reference gradient). Requires a working CUDA jaxlib on the target GPU.

Reproduced with a nonzero gradient error (`~770` max abs diff) on jax 0.11.0 on both a Volta (V100) and a Blackwell (RTX 5090) GPU; jax 0.10.2 does not reproduce it, consistent with this being a regression in a recent XLA snapshot.

```python
import jax
import jax.numpy as jnp
from jax.ad_checkpoint import checkpoint_name

jax.config.update("jax_default_matmul_precision", "highest")  # keep TF32 out of the comparison

BATCH, SEQ, WIDTH, LAYERS = 4, 256, 64, 8
MB = 2**20

OFFLOAD_CARRY = jax.checkpoint_policies.save_and_offload_only_these_names(
    names_which_can_be_saved=[],
    names_which_can_be_offloaded=["carry"],
    offload_src="device",
    offload_dst="pinned_host",
)


def layer(w, x):
    return jnp.tanh(x @ w)


def named_layer(w, x):
    # the only difference: the carry is named so the policy can act on it
    return layer(w, checkpoint_name(x, "carry"))


def stack(body):
    return lambda w, x: jax.lax.scan(lambda c, wi: (body(wi, c), None), x, w)[0]


def loss(body):
    return lambda w, x: jnp.sum(stack(body)(w, x) ** 2)


def report(name, body, w, x, reference=None):
    grad_fn = jax.jit(jax.grad(loss(body), argnums=(0, 1)))
    grads = grad_fn(w, x)
    stats = grad_fn.lower(w, x).compile().memory_analysis()
    delta = (
        None
        if reference is None
        else max(float(jnp.max(jnp.abs(a - b))) for a, b in zip(jax.tree.leaves(grads), jax.tree.leaves(reference)))
    )
    err = "--" if delta is None else f"{delta:.3e}"
    print(
        f"{name:26} device_temp={stats.temp_size_in_bytes / MB:7.1f} MB"
        f"  host_temp={stats.host_temp_size_in_bytes / MB:6.1f} MB"
        f"  max|grad - plain|={err}"
    )
    return grads, delta


def main():
    print(f"jax {jax.__version__}, device {jax.devices()[0].device_kind}")
    w = jax.random.normal(jax.random.key(0), (LAYERS, WIDTH, WIDTH)) * 0.3
    x = jax.random.normal(jax.random.key(1), (BATCH, SEQ, WIDTH))
    print(f"carry stack is {LAYERS * BATCH * SEQ * WIDTH * 4 / MB:.1f} MB\n")

    remat = lambda fn, policy: jax.checkpoint(fn, policy=policy, prevent_cse=False)

    plain, _ = report("plain scan", layer, w, x)
    report("remat nothing_saveable", remat(layer, jax.checkpoint_policies.nothing_saveable), w, x, plain)
    _, offload_err = report("remat offload 'carry'", remat(named_layer, OFFLOAD_CARRY), w, x, plain)

    if offload_err is not None and offload_err < 1e-5:
        print(
            "\nThe offload row matches plain scan, so this build is NOT affected:"
            "\neither XLA fixed it, or you are running a patched"
            "\nDynamicSliceAnnotator / AnalyzeDynamicSlice. Note the memory columns --"
            "\nthe residuals still went to host, they are just read in the right order now."
        )
    else:
        print(
            "\nExpected: the offload row matches plain scan (same math, the residual's"
            "\nmemory space is the only difference). Actual: it does not."
            "\n"
            "\nA nonzero error above is this script working as intended: it demonstrates the"
            "\nbug and deliberately uses the broken remat-policy route. Offloading a scan's"
            "\nresiduals correctly needs an explicit custom_vjp instead -- one that indexes"
            "\nthe stack by a carried value rather than the loop counter, which is what this"
            "\nrepo's learning/utils/scan.py::offload_scan does (exact to f32 rounding)."
        )


if __name__ == "__main__":
    main()
```

### CODE REVIEW

**Comment by zmelumian972** [comment]
I signed it!

**Comment by zmelumian972** [comment]
## Verification

Ran the repro script from the PR description on real hardware (Tesla V100-SXM2-16GB), comparing stock `jax/jaxlib 0.11.0` (PyPI release, unpatched) against a self-built `jaxlib` from this branch (`f5119f054c`, built after the fix landed).

| Build | jaxlib | `max|grad - plain|` for offloaded `carry` |
| --- | --- | --- |
| stock PyPI release (unpatched) | `0.11.0` | `7.695e+02` — bug reproduces |
| self-built from this branch | `0.11.0.dev0+selfbuilt` | `0.000e+00` — matches non-offloaded reference |

Same script, same seeds, same GPU — the only difference is the `AnalyzeDynamicSlice` fix. Also confirmed the diff is clean against `clang-format -style=file` (repo default, `BasedOnStyle: Google`) on both changed files.

**Comment by zmelumian972** [comment]
## Verified against a self-built jaxlib with the actual patch

Beyond the version correlation above, I built jaxlib + jax-cuda12-plugin + jax-cuda12-pjrt from source against jax v0.11.0's pinned XLA commit (`131bf41acb4650e4391a640c3f1859c1c86ad74b`) with this fix cherry-picked on top, and reran the repro on a V100:

```
jax 0.11.0, device Tesla V100-SXM2-16GB
carry stack is 2.0 MB

plain scan                 device_temp=   10.3 MB  host_temp=   0.0 MB  max|grad - plain|=--
remat nothing_saveable     device_temp=    6.8 MB  host_temp=   0.0 MB  max|grad - plain|=0.000e+00
remat offload 'carry'      device_temp=    4.8 MB  host_temp=   2.0 MB  max|grad - plain|=0.000e+00

NOT affected: offload row matches plain scan.
```

Host offload still genuinely happens (`host_temp=2.0 MB`), but the gradient now matches the plain scan exactly (`0.000e+00`), versus `7.695e+02` on an unpatched build. This confirms the fix resolves the bug, not just that it correlates with an XLA version bump.

**Comment by ezhulenev** [comment]
This change is too conservative and rejects many valid configs, I ported accidentally dropped support for cross-call-like induction variable trackign in https://github.com/openxla/xla/pull/46308. This used to work fine in V1 (now removed) version of the code. Thanks for reporting and the jax repro.

**Comment by akuegel** [comment]
@ezhulenev thanks for taking this over. @zmelumian972 I think we can close this PR? Eugene is fixing this issue in https://github.com/openxla/xla/pull/46308

**Comment by zmelumian972** [comment]
sure

**Comment by akuegel** [comment]
@zmelumian972 thank you for the bug report and your contribution. Closing this now and we will get Eugene's fix merged.
